File: 프로그래머스/2/43165. 타겟 넘버/타겟 넘버.py
from collections import deque
def solution(numbers, target):
    N=len(numbers)
    count=0
    def dfs(state,level):
        count=0
        # base-condition
        if level==N:
            return 1 if state==target else 0
        # do
        count+=dfs(state+numbers[level],level+1)
        count+=dfs(state-numbers[level],level+1)
        # undo
        
        return count
    

    return dfs(0,0)

File: 프로그래머스/2/43165. 타겟 넘버/타겟 넘버.java
class Solution {
    public int solution(int[] numbers, int target) {
        int answer = dfs(numbers,target,0,0);
        return answer;
    }
    
    private int dfs(int[] numbers,int target,int depth,int cur_num){
        int count=0;
        // base-condition
        if (depth==numbers.length){
            return target==cur_num?1:0;
        }
        // apply
        count+=dfs(numbers,target,depth+1,cur_num+numbers[depth]);
        count+=dfs(numbers,target,depth+1,cur_num-numbers[depth]);
        return count;
    }
}